cookbook/rule/add_custom_action_rule_engine.rst

```diff
@@ -0,0 +1,354 @@
+                    $result = str_replace('%%' . $attributeCode . '%%', $content, $result);
+                }
+
+                $this->propertySetter->setData(
+                    $product,
+                    $action->getField(),
+                    $result,
+                    $action->getOptions()
+                );
+            }
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function supports(ActionInterface $action)
+        {
+            return $action instanceof ProductPatternAction;
+        }
+    }
+
+Then we need to create the object that will handle the data.
+
+.. code-block:: php
+
+    #/src/Acme/Bundle/CustomBundle/Model/ProductPatternAction.php
+    <?php
+
+    namespace Acme\Bundle\CustomBundle\Model;
+
+    use Akeneo\Bundle\RuleEngineBundle\Model\ActionInterface;
+    use PimEnterprise\Component\CatalogRule\Model\ProductAddActionInterface;
+
+    class ProductPatternAction implements ActionInterface
+    {
+        const ACTION_TYPE = 'pattern';
+
+        /** @var string */
+        protected $field;
+
+        /** @var array */
+        protected $attributes = [];
+
+        /** @var string */
+        protected $pattern;
+
+        /** @var array */
+        protected $options = [];
+
+        /**
+         * {@inheritdoc}
+         */
+        public function getField()
+        {
+            return $this->field;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function setField($field)
+        {
+            $this->field = $field;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function getOptions()
+        {
+            return $this->options;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function setOptions(array $options = [])
+        {
+            $this->options = $options;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function getImpactedFields()
+        {
+            return [$this->getField()];
+        }
+
+        /**
+         * @return array
+         */
+        public function getAttributes()
+        {
+            return $this->attributes;
+        }
+
+        /**
+         * @return array
+         */
+        public function getPattern()
+        {
+            return $this->pattern;
+        }
+
+        /**
+         * @param array $attributes
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setAttributes(array $attributes = [])
+        {
+            $this->attributes = $attributes;
+        }
+
+        /**
+         * @param string $pattern
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setPattern($pattern)
+        {
+            $this->pattern = $pattern;
+        }
+    }
+
+We also need to create a denormalizer that will return our previous object that handles the data. It will convert the array into an object.
```

Comment:
This doesn't seem very clear to me and I had somes issues at first to explain this in trinings.
You should add why do we have an array > object denormalization? Where does this occur?